WEB: Replace os.path with pathlib.Path in pandas_web.py

iabhi4 opened the pull request:
Replaces `os.path` with `pathlib.Path` in `pandas_web.py`, as suggested by @datapythonista in [61578](https://github.com/pandas-dev/pandas/pull/61578). No functional changes, verified site generation remains correct

- [x] Ran pre-commit check

iabhi4 committed:
REF: Replace os.path with pathlib.Path in pandas_web.py

datapythonista commented on web/pandas_web.py:
```diff
@@ -442,16 +439,16 @@ def main(
             ) from e
 
-    config_fname = os.path.join(source_path, "config.yml")
+    config_fname = pathlib.Path(source_path) / "config.yml"
```
You are making source_path a pathlib twice. Ideally we want paths to be a pathlib since they are created. Not create them as str and convert when there is an operation. For everything that is a path, also config_path and others

datapythonista commented on web/pandas_web.py:
```diff
@@ -460,20 +457,19 @@ def main(
 
     for fname in get_source_files(source_path):
-        if os.path.normpath(fname) in context["main"]["ignore"]:
+        if str(pathlib.Path(fname)) in context["main"]["ignore"]:
```
I think there is an attribute for that, right?

iabhi4 replied:
Switched to` as_posix()` here, works cleanly with the `ignore` check now

datapythonista reviewed (commented):
Looks good, but we need to make everything that is a path a pathlib object when it's first created. Not create it as a string and cast it for every operation.

iabhi4 commented on web/pandas_web.py:
```diff
@@ -414,9 +411,9 @@ def get_source_files(source_path: str) -> typing.Generator[str, None, None]:
     Generate the list of files present in the source directory.
     """
     for root, dirs, fnames in os.walk(source_path):
```
Updated to create `Path` early in `main()`

iabhi4 commented on web/pandas_web.py:
```diff
@@ -394,7 +391,7 @@ def get_context(config_fname: str, **kwargs):
     with open(config_fname, encoding="utf-8") as f:
```
yes, changed it

iabhi4 commented:
Thanks for the review! updated to make sure all paths are created as `Path` early and used consistently. Removed redundant casts too.

iabhi4 commented on web/pandas_web.py:
```diff
@@ -100,20 +100,17 @@ def blog_add_posts(context):
         if context["blog"]["posts_path"]:
-            posts_path = os.path.join(
-                context["source_path"], *context["blog"]["posts_path"].split("/")
+            posts_path = pathlib.Path(context["source_path"]) / pathlib.Path(
```
simplified this to just `context["source_path"] / context["blog"]["posts_path"]`

datapythonista reviewed (commented):
Looks good. Added few comments, but very nice clean up

iabhi4 commented on web/pandas_web.py:
```diff
@@ -409,14 +404,14 @@ def get_context(config_fname: str, **kwargs):
     """
-    for root, dirs, fnames in os.walk(source_path):
-        root_rel_path = os.path.relpath(root, source_path)
+    for root, dirs, fnames in os.walk(str(source_path)):
```
Using` Path.rglob()` now

iabhi4 commented on web/pandas_web.py:
```diff
@@ -441,17 +436,20 @@ def main(
     """
+    source_path = pathlib.Path(source_path)
+    target_path = pathlib.Path(target_path)
+
```
You are right, Now main() receives a `Path` obj

iabhi4 commented on web/pandas_web.py:
```diff
@@ -460,23 +458,22 @@ def main(
             if extension == ".md":
-                if "pdeps/" in fname:
+                if "pdeps" in fname.parts:
                     from markdown.extensions.toc import TocExtension
```
Updated

iabhi4 commented on web/pandas_web.py:
```diff
@@ -503,17 +500,13 @@ def main(
-            shutil.copy(
-                os.path.join(source_path, fname), os.path.join(target_path, dirname)
-            )
+            shutil.copy(source_path / fname, target_path / dirname / fname.name)
```
Changed it to target_path / fname, Works fine

iabhi4 commented:
Addressed all the latest comments, switched to `Path.rglob`, cleaned up `base_url`, and ensured consistent `Path` usage. Let me know if anything else looks off, I also retested on local after the changes

datapythonista committed:
Merge branch 'main' into refactor-pathlib-clean

datapythonista committed:
Merge branch 'main' into refactor-pathlib-clean
